File: taurus-protect-sdk-python/taurus_protect/services/transaction_service.py
"""Transaction service for Taurus-PROTECT SDK."""

from __future__ import annotations

from datetime import datetime
from typing import TYPE_CHECKING, Any, List, Optional, Tuple

from taurus_protect._internal.openapi.exceptions import ApiException
from taurus_protect.errors import NotFoundError
from taurus_protect.mappers.transaction import map_transaction, map_transactions
from taurus_protect.models.pagination import (
    PLUS_ROWS,
    Pagination,
    offset_pagination,
    offset_query,
    parse_count,
    resolve_offset,
    resolve_page_size,
)
from taurus_protect.models.transaction import Transaction, TransactionExport
from taurus_protect.services._base import BaseService

if TYPE_CHECKING:
    from taurus_protect._internal.openapi.api.transactions_api import TransactionsApi


class TransactionService(BaseService):
    """
    Service for retrieving and exporting blockchain transactions.

    Transactions represent the movement of cryptocurrency on the blockchain
    and can be either incoming (received) or outgoing (sent).

    Example:
        >>> # Get recent transactions
        >>> transactions, pagination = client.transactions.list(currency="ETH", limit=100)
        >>> for tx in transactions:
        ...     print(f"{tx.tx_hash}: {tx.amount} {tx.currency}")
        ...
        >>> # Get a transaction by hash
        >>> tx = client.transactions.get_by_hash("0x1234...")
    """

    def __init__(
        self,
        api_client: Any,
        transactions_api: "TransactionsApi",
    ) -> None:
        """
        Initialize the transaction service.

        Args:
            api_client: The OpenAPI client instance.
            transactions_api: The transactions API instance.
        """
        super().__init__(api_client)
        self._api = transactions_api

    def get(self, transaction_id: int) -> Transaction:
        """
        Get a single transaction by ID.

        Args:
            transaction_id: The transaction ID.

        Returns:
            The transaction.

        Raises:
            NotFoundError: If the transaction is not found.
            APIError: If the API call fails.
        """
        if transaction_id <= 0:
            raise ValueError("transaction_id must be positive")

        try:
            reply = self._api.transaction_service_get_transactions(
                currency=None,
                direction=None,
                query=None,
                limit="1",
                offset="0",
                var_from=None,
                to=None,
                transaction_ids=None,
                type=None,
                source=None,
                destination=None,
                ids=[str(transaction_id)],
                blockchain=None,
                network=None,
                from_block_number=None,
                to_block_number=None,
                hashes=None,
                address=None,
                amount_above=None,
                exclude_unknown_source_destination=None,
                customer_id=None,
            )

            result = reply.result
            if not result:
                raise NotFoundError(f"Transaction with id '{transaction_id}' not found")

            return map_transaction(result[0])
        except NotFoundError:
            raise
        except Exception as e:
            if isinstance(e, ApiException):
                raise self._handle_error(e)
            raise

    def get_by_hash(self, tx_hash: str) -> Transaction:
        """
        Get a transaction by its blockchain hash.

        Args:
            tx_hash: The transaction hash.

        Returns:
            The transaction.

        Raises:
            NotFoundError: If the transaction is not found.
            APIError: If the API call fails.
        """
        self._validate_required(tx_hash, "tx_hash")

        try:
            reply = self._api.transaction_service_get_transactions(
                currency=None,
                direction=None,
                query=None,
                limit="1",
                offset="0",
                var_from=None,
                to=None,
                transaction_ids=None,
                type=None,
                source=None,
                destination=None,
                ids=None,
                blockchain=None,
                network=None,
                from_block_number=None,
                to_block_number=None,
                hashes=[tx_hash],
                address=None,
                amount_above=None,
                exclude_unknown_source_destination=None,
                customer_id=None,
            )

            result = reply.result
            if not result:
                raise NotFoundError(f"Transaction with hash '{tx_hash}' not found")

            return map_transaction(result[0])
        except NotFoundError:
            raise
        except Exception as e:
            if isinstance(e, ApiException):
                raise self._handle_error(e)
            raise

    def list(
        self,
        from_date: Optional[datetime] = None,
        to_date: Optional[datetime] = None,
        currency: Optional[str] = None,
        direction: Optional[str] = None,
        limit: Optional[int] = None,
        offset: Optional[int] = None,
        blockchain: Optional[str] = None,
        network: Optional[str] = None,
    ) -> Tuple[List[Transaction], Pagination]:
        """
        List transactions with filtering, one page at a time.

        Args:
            from_date: Filter transactions after this date.
            to_date: Filter transactions before this date.
            currency: Filter by currency ID or symbol.
            direction: Filter by direction ("incoming" or "outgoing").
            limit: Page size (default 20, max 100).
            offset: Number of transactions to skip; pass ``pagination.next_offset``.
            blockchain: Filter by blockchain (e.g. "ETH").
            network: Filter by network (e.g. "mainnet").

        Returns:
            Tuple of (transactions, pagination). The server's total can be an upper
            bound, so a trailing page may come back empty with ``has_more`` false.

        Raises:
            ValueError: If limit or offset are invalid.
            APIError: If the API call fails.
        """
        return self._list(
            limit,
            offset,
            currency=currency,
            direction=direction,
            var_from=from_date,
            to=to_date,
            blockchain=blockchain,
            network=network,
        )

    def list_by_address(
        self,
        address: str,
        limit: Optional[int] = None,
        offset: Optional[int] = None,
    ) -> Tuple[List[Transaction], Pagination]:
        """
        List transactions for a specific blockchain address, one page at a time.

        Args:
            address: The blockchain address.
            limit: Page size (default 20, max 100).
            offset: Number of transactions to skip; pass ``pagination.next_offset``.

        Returns:
            Tuple of (transactions, pagination).

        Raises:
            ValueError: If address is empty or limit/offset are invalid.
            APIError: If the API call fails.
        """
        self._validate_required(address, "address")
        return self._list(limit, offset, address=address)

    def _list(
        self, limit: Optional[int], offset: Optional[int], **filters: Any
    ) -> Tuple[List[Transaction], Pagination]:
        page_size = resolve_page_size(limit, "limit")
        start = resolve_offset(offset)

        try:
            reply = self._api.transaction_service_get_transactions(
                **offset_query(page_size, start), **filters
            )

            rows = reply.result or []
            transactions = map_transactions(rows)
            pagination = offset_pagination(
                PLUS_ROWS,
                limit=page_size,
                offset=start,
                served_rows=len(rows),
                total_items=reply.total_items,
                excluded=len(rows) - len(transactions),
            )
            return transactions, pagination
        except Exception as e:
            if isinstance(e, ApiException):
                raise self._handle_error(e)
            raise

    def export(
        self,
        from_date: Optional[datetime] = None,
        to_date: Optional[datetime] = None,
        currency: Optional[str] = None,
        direction: Optional[str] = None,
        limit: Optional[int] = None,
        blockchain: Optional[str] = None,
        network: Optional[str] = None,
        format: Optional[str] = None,
    ) -> TransactionExport:
        """
        Export transactions in one reply.

        The export cannot page: the server always exports from the first matching row,
        so there is no offset. When ``total_items`` exceeds the rows exported, raise the
        limit or narrow the filters.

        Args:
            from_date: Filter transactions after this date.
            to_date: Filter transactions before this date.
            currency: Filter by currency ID or symbol.
            direction: Filter by direction ("incoming" or "outgoing").
            limit: Maximum rows to export (default 20, no SDK maximum).
            blockchain: Filter by blockchain (e.g. "ETH").
            network: Filter by network (e.g. "mainnet").
            format: Export format ("json", "csv", "csv_simple"); the server defaults
                to JSON.

        Returns:
            The exported text and the server's count of matching transactions.

        Raises:
            ValueError: If limit is negative.
            APIError: If the API call fails.
        """
        size = resolve_page_size(limit, "limit", maximum=None)

        try:
            reply = self._api.transaction_service_export_transactions(
                currency=currency,
                direction=direction,
                var_from=from_date,
                to=to_date,
                format=format,
                blockchain=blockchain,
                network=network,
                limit=str(size),
            )

            return TransactionExport(
                content=reply.result or "",
                total_items=parse_count(reply.total_items, "totalItems"),
            )
        except Exception as e:
            if isinstance(e, ApiException):
                raise self._handle_error(e)
            raise

    def export_csv(
        self,
        from_date: Optional[datetime] = None,
        to_date: Optional[datetime] = None,
        currency: Optional[str] = None,
        direction: Optional[str] = None,
        limit: Optional[int] = None,
        blockchain: Optional[str] = None,
        network: Optional[str] = None,
    ) -> TransactionExport:
        """
        Export transactions as CSV: :meth:`export` with ``format="csv"``.

        Args:
            from_date: Filter transactions after this date.
            to_date: Filter transactions before this date.
            currency: Filter by currency ID or symbol.
            direction: Filter by direction ("incoming" or "outgoing").
            limit: Maximum rows to export (default 20, no SDK maximum).
            blockchain: Filter by blockchain (e.g. "ETH").
            network: Filter by network (e.g. "mainnet").

        Returns:
            The CSV text and the server's count of matching transactions.

        Raises:
            ValueError: If limit is negative.
            APIError: If the API call fails.
        """
        return self.export(
            from_date=from_date,
            to_date=to_date,
            currency=currency,
            direction=direction,
            limit=limit,
            blockchain=blockchain,
            network=network,
            format="csv",
        )

File: taurus-protect-sdk-typescript/src/services/transaction-service.ts
/**
 * Transaction service for Taurus-PROTECT SDK.
 *
 * Provides methods for retrieving and filtering blockchain transactions.
 */

import { NotFoundError, ValidationError } from "../errors";
import type { TransactionsApi } from "../internal/openapi/apis/TransactionsApi";
import { transactionsFromDto } from "../mappers/transaction";
import {
  buildOffsetPagination,
  offsetRequest,
  parseReplyCount,
  resolvePageSize,
  type OffsetPageOptions,
  type PaginatedResult,
} from "../models/pagination";
import type {
  ExportTransactionsOptions,
  ExportTransactionsResult,
  ListTransactionsOptions,
  Transaction,
} from "../models/transaction";
import { BaseService } from "./base";
import { offsetQuery } from "./paging";

/**
 * Service for retrieving and filtering blockchain transactions.
 *
 * Transactions represent the movement of cryptocurrency on the blockchain
 * and can be either incoming (received) or outgoing (sent).
 *
 * @example
 * ```typescript
 * // Get a single transaction by ID
 * const tx = await transactionService.get("12345");
 * console.log(`${tx.hash}: ${tx.amount} ${tx.currency}`);
 *
 * // List transactions with filters
 * const result = await transactionService.list({
 *   currency: "ETH",
 *   direction: "incoming",
 *   limit: 50,
 * });
 * for (const tx of result.items) {
 *   console.log(`${tx.hash}: ${tx.amount}`);
 * }
 *
 * // List transactions by request ID
 * const txs = await transactionService.listByRequest("req-123");
 * ```
 */
export class TransactionService extends BaseService {
  private readonly transactionsApi: TransactionsApi;

  /**
   * Creates a new TransactionService instance.
   *
   * @param transactionsApi - The TransactionsApi instance from the OpenAPI client
   */
  constructor(transactionsApi: TransactionsApi) {
    super();
    this.transactionsApi = transactionsApi;
  }

  /**
   * Gets a transaction by ID.
   *
   * @param transactionId - The transaction ID to retrieve
   * @returns The transaction
   * @throws {@link ValidationError} If transactionId is invalid
   * @throws {@link NotFoundError} If transaction not found
   * @throws {@link APIError} If API request fails
   *
   * @example
   * ```typescript
   * const tx = await transactionService.get("12345");
   * console.log(`Hash: ${tx.hash}, Amount: ${tx.amount}`);
   * ```
   */
  async get(transactionId: string): Promise<Transaction> {
    if (!transactionId || transactionId.trim() === "") {
      throw new ValidationError("transactionId is required");
    }

    return this.execute(async () => {
      const response = await this.transactionsApi.transactionServiceGetTransactions({
        ids: [transactionId],
        limit: "1",
      });

      const transactions = transactionsFromDto(response.result);
      if (transactions.length === 0) {
        throw new NotFoundError(`Transaction ${transactionId} not found`);
      }

      return transactions[0];
    });
  }

  /**
   * Gets a transaction by its blockchain hash.
   *
   * @param txHash - The blockchain transaction hash
   * @returns The transaction
   * @throws {@link ValidationError} If txHash is invalid
   * @throws {@link NotFoundError} If transaction not found
   * @throws {@link APIError} If API request fails
   *
   * @example
   * ```typescript
   * const tx = await transactionService.getByHash("0x1234...");
   * console.log(`ID: ${tx.id}, Amount: ${tx.amount}`);
   * ```
   */
  async getByHash(txHash: string): Promise<Transaction> {
    if (!txHash || txHash.trim() === "") {
      throw new ValidationError("txHash is required");
    }

    return this.execute(async () => {
      const response = await this.transactionsApi.transactionServiceGetTransactions({
        hashes: [txHash],
        limit: "1",
      });

      const transactions = transactionsFromDto(response.result);
      if (transactions.length === 0) {
        throw new NotFoundError(`Transaction with hash ${txHash} not found`);
      }

      return transactions[0];
    });
  }

  /**
   * Lists transactions with pagination and optional filtering.
   *
   * @param options - Optional filtering and pagination options
   * @returns Paginated result containing transactions and pagination info
   * @throws {@link ValidationError} If limit or offset are invalid
   * @throws {@link APIError} If API request fails
   *
   * @example
   * ```typescript
   * // List first 50 transactions
   * const result = await transactionService.list({ limit: 50, offset: 0 });
   * console.log(`Found ${result.pagination.totalItems} transactions`);
   *
   * // List with filters
   * const filtered = await transactionService.list({
   *   currency: "ETH",
   *   direction: "incoming",
   *   blockchain: "ETH",
   *   network: "mainnet",
   * });
   * ```
   */
  async list(
    options?: ListTransactionsOptions
  ): Promise<PaginatedResult<Transaction>> {
    const page = offsetRequest(options);

    return this.execute(async () => {
      const response = await this.transactionsApi.transactionServiceGetTransactions({
        currency: options?.currency,
        direction: options?.direction,
        query: options?.query,
        ...offsetQuery(page),
        from: options?.fromDate,
        to: options?.toDate,
        type: options?.type,
        source: options?.source,
        destination: options?.destination,
        ids: options?.ids,
        blockchain: options?.blockchain,
        network: options?.network,
        fromBlockNumber: options?.fromBlockNumber,
        toBlockNumber: options?.toBlockNumber,
        hashes: options?.hashes,
        address: options?.address,
        amountAbove: options?.amountAbove,
        excludeUnknownSourceDestination: options?.excludeUnknownSourceDestination,
        customerId: options?.customerId,
      });

      const rows = response.result ?? [];
      return {
        items: transactionsFromDto(rows),
        pagination: buildOffsetPagination("plus_rows", page, response, rows.length),
      };
    });
  }

  /**
   * Lists transactions associated with a specific request ID.
   *
   * @param requestId - The request ID to filter by
   * @param options - Optional pagination options
   * @returns Paginated result containing transactions
   * @throws {@link ValidationError} If requestId is invalid
   * @throws {@link APIError} If API request fails
   *
   * @example
   * ```typescript
   * const result = await transactionService.listByRequest("req-123");
   * for (const tx of result.items) {
   *   console.log(`${tx.hash}: ${tx.status}`);
   * }
   * ```
   */
  async listByRequest(
    requestId: string,
    options?: OffsetPageOptions
  ): Promise<PaginatedResult<Transaction>> {
    if (!requestId || requestId.trim() === "") {
      throw new ValidationError("requestId is required");
    }
    const page = offsetRequest(options);

    return this.execute(async () => {
      // Note: The API doesn't have a direct requestId filter, so we use transactionIds
      // which maps to the internal transaction IDs associated with requests
      const response = await this.transactionsApi.transactionServiceGetTransactions({
        transactionIds: [requestId],
        ...offsetQuery(page),
      });

      const rows = response.result ?? [];
      return {
        items: transactionsFromDto(rows),
        pagination: buildOffsetPagination("plus_rows", page, response, rows.length),
      };
    });
  }

  /**
   * Lists transactions for a specific blockchain address.
   *
   * @param address - The blockchain address
   * @param options - Optional pagination options
   * @returns Paginated result containing transactions
   * @throws {@link ValidationError} If address is invalid
   * @throws {@link APIError} If API request fails
   *
   * @example
   * ```typescript
   * const result = await transactionService.listByAddress("0x1234...");
   * for (const tx of result.items) {
   *   console.log(`${tx.hash}: ${tx.amount}`);
   * }
   * ```
   */
  async listByAddress(
    address: string,
    options?: OffsetPageOptions
  ): Promise<PaginatedResult<Transaction>> {
    if (!address || address.trim() === "") {
      throw new ValidationError("address is required");
    }
    const page = offsetRequest(options);

    return this.execute(async () => {
      const response = await this.transactionsApi.transactionServiceGetTransactions({
        address,
        ...offsetQuery(page),
      });

      const rows = response.result ?? [];
      return {
        items: transactionsFromDto(rows),
        pagination: buildOffsetPagination("plus_rows", page, response, rows.length),
      };
    });
  }

  /**
   * Exports transactions as a formatted string (JSON or CSV).
   *
   * The export cannot page — the server always starts at the first matching row — so it
   * takes a `limit` (default 20, no SDK maximum) and no offset. The result's `totalItems`
   * tells a truncated export apart from a complete one.
   *
   * @param options - Filters, format and `limit`
   * @returns The export payload and the server's total
   * @throws {@link ValidationError} If the limit is negative
   * @throws {@link PaginationError} If the reply's total is not a canonical count
   * @throws {@link APIError} If API request fails
   *
   * @example
   * ```typescript
   * const exported = await transactionService.exportTransactions({
   *   currency: 'ETH',
   *   fromDate: new Date('2024-01-01'),
   *   format: 'csv',
   *   limit: 1000,
   * });
   * console.log(exported.data);
   * console.log(`${exported.totalItems} transactions match`);
   * ```
   */
  async exportTransactions(
    options?: ExportTransactionsOptions
  ): Promise<ExportTransactionsResult> {
    const limit = resolvePageSize(options?.limit, "limit", "export");

    return this.execute(async () => {
      const response = await this.transactionsApi.transactionServiceExportTransactions({
        from: options?.fromDate,
        to: options?.toDate,
        currency: options?.currency,
        direction: options?.direction,
        limit: String(limit),
        format: options?.format,
        blockchain: options?.blockchain,
        network: options?.network,
        query: options?.query,
        address: options?.address,
      });
      return {
        data: response.result ?? "",
        totalItems: parseReplyCount(response.totalItems, "totalItems"),
      };
    });
  }
}
